PR: https://github.com/odoo/odoo/pull/222261

From: 2ead34e974f55dadf791ff6b2924edbf43442b34
From: David Monnom
Date: 2025-09-08 15:26:52

Breaking data model changes scores: del:5 + add:0, change matches:
-    sequence_id = fields.Many2one('ir.sequence', string='Order IDs Sequence', readonly=True,
-    sequence_line_id = fields.Many2one('ir.sequence', string='Order Line IDs Sequence', readonly=True,
-    order_seq_id = fields.Many2one('ir.sequence', string='Order Sequence', readonly=True, copy=False, help='Used to generate the OOOO part of the pos_reference field of the pos.order model.')
-    login_number_seq_id = fields.Many2one('ir.sequence', string='Login Number Sequence', readonly=True, copy=False, help='Determines the number of times the UI is opened. It is used as proxy to the identity of the device where the UI is opened. And as such, it is the LL part of the pos_reference field of the pos.order model.')
-    pos_sequence_id = fields.Many2one(related='pos_config_id.sequence_id')

Total Changes: 196

[IMP] point_of_sale, pos_*: Revamp sequence management in POS

*: l10n_co_pos, l10n_fr_pos_cert, pos_restaurant, pos_self_order

Sequence management has been reworked. Here is how it now looks:
* `order_seq_id`: this sequence is used for all orders in the database.
* `order_backend_seq_id`: this sequence is used for all orders that are
created from something other than a Point of Sale, for example self-order
or a refund from the backend.
* `order_line_seq_id`: this sequence is used for all order lines in
the database.
* `device_seq_id`: this sequence is used to identify all devices that
will load the Point of Sale (separated by pos_config).

`pos_reference` are now created with this format: `000-0-000000`
* First part: Two-digit year
* Second part: PoS ID
* Third part: incremented sequence, minimum 6 digits (may be more).
This sequence is linked to the identity of the first part.

Taskid: 5004675
Part-of: odoo/odoo#222261
Related: odoo/enterprise#91962
Related: odoo/upgrade#8197
Signed-off-by: Stéphane Vanmeerhaeghe (stva) <stva@odoo.com>

================================= pseudo patch: =================================

--- a/addons/point_of_sale/models/ir_sequence.py
+++ b/addons/point_of_sale/models/ir_sequence.py
@@ -7,10 +7,15 @@ class IrSequence(models.Model):
 
     @api.ondelete(at_uninstall=False)
     def _unlink_sequence(self):
-        sessions = self.env['pos.session'].search(domain=['|', ('login_number_seq_id', 'in', self.ids), ('order_seq_id', 'in', self.ids)])
-        acive_sessions = sessions.filtered(lambda s: s.state != 'closed')
-        if acive_sessions:
+        configs = self.env['pos.config'].search(domain=[
+            '|', '|', '|',
+            ('order_seq_id', 'in', self.ids),
+            ('order_line_seq_id', 'in', self.ids),
+            ('device_seq_id', 'in', self.ids),
+            ('order_backend_seq_id', 'in', self.ids)
+        ])
+        if len(configs):
             raise UserError(_(
-                "You cannot delete a sequence used in an active session: %s",
-                acive_sessions.login_number_seq_id.mapped('name')
+                "You cannot delete a sequence used in an active POS config: %s",
+                configs.order_seq_id.mapped('name')
             ))

--- a/addons/point_of_sale/models/pos_config.py
+++ b/addons/point_of_sale/models/pos_config.py
@@ -95,6 +95,10 @@ class PosConfig(models.Model):
         help="Accounting journal used to create invoices.",
         default=_default_invoice_journal)
     currency_id = fields.Many2one('res.currency', compute='_compute_currency', store=True, compute_sudo=True, string="Currency")
+    order_seq_id = fields.Many2one('ir.sequence', string='Order Sequence', readonly=True, copy=False)
+    order_backend_seq_id = fields.Many2one('ir.sequence', string='Order Backend Sequence', readonly=True, copy=False)
+    order_line_seq_id = fields.Many2one('ir.sequence', string='Order Line Sequence', readonly=True, copy=False)
+    device_seq_id = fields.Many2one('ir.sequence', string='Device Sequence', readonly=True, copy=False)
     iface_cashdrawer = fields.Boolean(string='Cashdrawer', help="Automatically open the cashdrawer.")
     iface_electronic_scale = fields.Boolean(string='Electronic Scale', help="Enables Electronic Scale integration.")
     iface_print_via_proxy = fields.Boolean(string='Print via Proxy', help="Bypass browser printing and prints via the hardware proxy.")
@@ -124,12 +128,6 @@ class PosConfig(models.Model):
     active = fields.Boolean(default=True)
     uuid = fields.Char(readonly=True, default=lambda self: str(uuid4()), copy=False,
         help='A globally unique identifier for this pos configuration, used to prevent conflicts in client-generated data.')
-    sequence_id = fields.Many2one('ir.sequence', string='Order IDs Sequence', readonly=True,
-        help="This sequence is automatically created by Odoo but you can change it "
-        "to customize the reference numbers of your orders.", copy=False, ondelete='restrict')
-    sequence_line_id = fields.Many2one('ir.sequence', string='Order Line IDs Sequence', readonly=True,
-        help="This sequence is automatically created by Odoo but you can change it "
-        "to customize the reference numbers of your orders lines.", copy=False)
     session_ids = fields.One2many('pos.session', 'config_id', string='Sessions')
     current_session_id = fields.Many2one('pos.session', compute='_compute_current_session', string="Current Session")
     current_session_state = fields.Char(compute='_compute_current_session')
@@ -210,7 +208,13 @@ class PosConfig(models.Model):
         store=True, help="These payment methods will be available for fast payment", readonly=False)
     statistics_for_current_session = fields.Json(string="Session Statistics", compute="_compute_statistics_for_session")
 
-    def notify_synchronisation(self, session_id, login_number, records={}):
+    def _get_next_order_refs(self, device_identifier='0'):
+        next_number = self.order_backend_seq_id._next()
+        year_2_digits = str(datetime.now().year)[-2:]
+        tracking_number = f"{int(next_number) % 1000}"
+        return f"{year_2_digits}{device_identifier}-{self.id}-{next_number}", tracking_number
+
+    def notify_synchronisation(self, session_id, device_identifier, records={}):
         self.ensure_one()
         static_records = {}
 
@@ -221,7 +225,7 @@ class PosConfig(models.Model):
         self._notify('SYNCHRONISATION', {
             'static_records': static_records,
             'session_id': session_id,
-            'login_number': login_number,
+            'device_identifier': device_identifier,
             'records': records
         })
 
@@ -515,9 +519,6 @@ class PosConfig(models.Model):
         if not self.env.is_admin() and {'is_header_or_footer', 'receipt_header', 'receipt_footer'} & values.keys():
             raise AccessError(_('Only administrators can edit receipt headers and footers'))
 
-    def _config_sequence_implementation(self):
-        return 'standard'
-
     @api.model_create_multi
     def create(self, vals_list):
         if not self._default_warehouse_id():
@@ -527,27 +528,56 @@ class PosConfig(models.Model):
             })
         for vals in vals_list:
             self._check_header_footer(vals)
-            IrSequence = self.env['ir.sequence'].sudo()
-            val = {
-                'name': _('POS Order %s', vals['name']),
-                'padding': 4,
-                'prefix': "%s/" % vals['name'],
-                'code': "pos.order",
-                'company_id': vals.get('company_id', False),
-                'implementation': self._config_sequence_implementation(),
-            }
-            # force sequence_id field to new pos.order sequence
-            vals['sequence_id'] = IrSequence.create(val).id
 
-            val.update(name=_('POS order line %s', vals['name']), code='pos.order.line')
-            vals['sequence_line_id'] = IrSequence.create(val).id
         pos_configs = super().create(vals_list)
+        pos_configs._create_sequences()
         pos_configs.sudo()._check_modules_to_install()
         pos_configs.sudo()._check_groups_implied()
         pos_configs._update_preparation_printers_menuitem_visibility()
         # If you plan to add something after this, use a new environment. The one above is no longer valid after the modules install.
         return pos_configs
 
+    def _create_sequences(self):
+        for pos_config in self:
+            sequence_vals = {
+                'padding': 6,
+                'code': "pos.order",
+                'company_id': pos_config.company_id.id,
+                'implementation': 'no_gap',
+            }
+
+            # Create sequences for all orders
+            pos_config.order_seq_id = self.env['ir.sequence'].sudo().create({
+                **sequence_vals,
+                'name': _('POS order from config #%s', pos_config.id),
+            })
+
+            # Create sequences for order that are created from self ore backend
+            pos_config.order_backend_seq_id = self.env['ir.sequence'].sudo().create({
+                **sequence_vals,
+                'name': _('POS order backend from config #%s', pos_config.id),
+            })
+
+            # Create sequences for all order lines
+            pos_config.order_line_seq_id = self.env['ir.sequence'].sudo().create({
+                **sequence_vals,
+                'name': _('POS order line from config #%s', pos_config.id),
+            })
+
+            # Create sequences for devices
+            pos_config.device_seq_id = self.env['ir.sequence'].sudo().create({
+                **sequence_vals,
+                'name': _('POS device from config #%s', pos_config.id),
+                'padding': 0,
+            })
+
+    def register_new_device_identifier(self):
+        self.ensure_one()
+        identifier = self.device_seq_id._next()
+        return {
+            'device_identifier': identifier,
+        }
+
     def _reset_default_on_vals(self, vals):
         if 'tip_product_id' in vals and not vals['tip_product_id'] and 'iface_tipproduct' in vals and vals['iface_tipproduct']:
             default_product = self.env.ref('point_of_sale.product_product_tip', False)
@@ -662,7 +692,7 @@ class PosConfig(models.Model):
 
     def unlink(self):
         # Delete the pos.config records first then delete the sequences linked to them
-        sequences_to_delete = self.sequence_id | self.sequence_line_id
+        sequences_to_delete = self.order_line_seq_id | self.device_seq_id
         res = super(PosConfig, self).unlink()
         sequences_to_delete.unlink()
         return res

--- a/addons/point_of_sale/models/pos_order.py
+++ b/addons/point_of_sale/models/pos_order.py
@@ -72,6 +72,9 @@ class PosOrder(models.Model):
         if pos_session.state == 'closing_control' or pos_session.state == 'closed':
             order['session_id'] = self._get_valid_session(order).id
 
+        if not order.get('source'):
+            order['source'] = 'pos'
+
         if order.get('partner_id'):
             partner_id = self.env['res.partner'].browse(order['partner_id'])
             if not partner_id.exists():
@@ -141,8 +144,11 @@ class PosOrder(models.Model):
             self._create_order_picking()
             self._compute_total_cost_in_real_time()
 
-        if self.to_invoice and self.state == 'paid':
+        if self.to_invoice and self.state == 'paid' and self.config_id.invoice_journal_id:
             self._generate_pos_order_invoice()
+        elif not self.config_id.invoice_journal_id:
+            _logger.warning('Trying to create an invoice without any journal configured')
+            raise UserError(_('No invoice journal configured for this POS session.'))
 
         return self.id
 
@@ -316,16 +322,7 @@ class PosOrder(models.Model):
     general_customer_note = fields.Text(string='General Customer Note')
     internal_note = fields.Text(string='Internal Note')
     nb_print = fields.Integer(string='Number of Print', readonly=True, copy=False, default=0)
-    pos_reference = fields.Char(string='Receipt Number', readonly=True, copy=False, index=True, help="""
-        Human readable reference for this order.
-            * Format: YYLL-SSS-FOOOO
-            * YY is the year
-            * LL is the login number (proxy to the device)
-            * SSS is the session id
-            * F is 1 or 0 (1 if sequence number is generated from client)
-            * OOOO is the sequence number.
-    """
-    )
+    pos_reference = fields.Char(string='Receipt Number', readonly=True, copy=False, index=True)
     sale_journal = fields.Many2one('account.journal', related='session_id.config_id.journal_id', string='Sales Journal', store=True, readonly=True, ondelete='restrict')
     fiscal_position_id = fields.Many2one(
         comodel_name='account.fiscal.position', string='Fiscal Position',
@@ -361,6 +358,7 @@ class PosOrder(models.Model):
         string="Reversal Account Moves",
         help="List of account moves created when this POS order was reversed and invoiced after session close."
     )
+    source = fields.Selection(string="Origin", selection=[('pos', 'Point of Sale')], default='pos')
 
     _unique_uuid = models.Constraint('unique (uuid)', 'An order with this uuid already exists')
 
@@ -539,11 +537,15 @@ class PosOrder(models.Model):
         values.setdefault('pricelist_id', session.config_id.pricelist_id.id)
         values.setdefault('fiscal_position_id', session.config_id.default_fiscal_position_id.id)
         values.setdefault('company_id', session.config_id.company_id.id)
-        if values.get('tracking_number') is None and values.get('pos_reference') is None and values.get('sequence_number') is None:
-            pos_reference, sequence_number, tracking_number = session.get_next_order_refs()
-            values['pos_reference'] = pos_reference
-            values['sequence_number'] = sequence_number
+
+        if not values.get('pos_reference'):
+            reference, tracking_number = session.config_id._get_next_order_refs()
+            values['pos_reference'] = reference
             values['tracking_number'] = tracking_number
+
+        if not values.get('sequence_number'):
+            values['sequence_number'] = session.config_id.order_seq_id._next()  # Some localization needs orders to have a sequence number
+
         return values
 
     def write(self, vals):
@@ -648,7 +650,8 @@ class PosOrder(models.Model):
         if self.refunded_order_id.exists():
             return _('%(refunded_order)s REFUND', refunded_order=self.refunded_order_id.name)
         else:
-            return session.config_id.sequence_id._next()
+            last_reference_part = self.pos_reference.split('-')[-1]
+            return f"{session.config_id.name} - {last_reference_part}"
 
     def action_stock_picking(self):
         self.ensure_one()
@@ -1160,7 +1163,7 @@ class PosOrder(models.Model):
         for order in pos_order_ids:
             order._ensure_access_token()
             if not self.env.context.get('preparation'):
-                order.config_id.notify_synchronisation(order.config_id.current_session_id.id, self.env.context.get('login_number', 0))
+                order.config_id.notify_synchronisation(order.config_id.current_session_id.id, self.env.context.get('device_identifier', 0))
 
         _logger.info("PoS synchronisation #%d finished", sync_token)
         return pos_order_ids.read_pos_data(orders, config)
@@ -1221,7 +1224,7 @@ class PosOrder(models.Model):
 
     def _prepare_refund_values(self, current_session):
         self.ensure_one()
-        pos_reference, sequence_number, tracking_number = current_session.get_next_order_refs()
+        pos_reference, tracking_number = current_session.config_id._get_next_order_refs()
         return {
             'name': _('%(name)s REFUND', name=self.name),
             'session_id': current_session.id,
@@ -1231,7 +1234,6 @@ class PosOrder(models.Model):
             'amount_paid': 0,
             'is_total_cost_computed': False,
             'is_refund': True,
-            'sequence_number': sequence_number,
             'tracking_number': tracking_number,
         }
 
@@ -1520,8 +1522,8 @@ class PosOrderLine(models.Model):
             if order and order.exists() and not vals.get('name'):
                 # set name based on the sequence specified on the config
                 config = order.session_id.config_id
-                if config.sequence_line_id:
-                    vals['name'] = config.sequence_line_id._next()
+                if config.order_line_seq_id:
+                    vals['name'] = config.order_line_seq_id._next()
             if not vals.get('name'):
                 # fallback on any pos.order sequence
                 vals['name'] = self.env['ir.sequence'].next_by_code('pos.order.line')

--- a/addons/point_of_sale/models/pos_session.py
+++ b/addons/point_of_sale/models/pos_session.py
@@ -50,9 +50,6 @@ class PosSession(models.Model):
         required=True, readonly=True,
         index=True, copy=False, default='opening_control')
 
-    order_seq_id = fields.Many2one('ir.sequence', string='Order Sequence', readonly=True, copy=False, help='Used to generate the OOOO part of the pos_reference field of the pos.order model.')
-    login_number_seq_id = fields.Many2one('ir.sequence', string='Login Number Sequence', readonly=True, copy=False, help='Determines the number of times the UI is opened. It is used as proxy to the identity of the device where the UI is opened. And as such, it is the LL part of the pos_reference field of the pos.order model.')
-
     opening_notes = fields.Text(string="Opening Notes")
     closing_notes = fields.Text(string="Closing Notes")
     cash_control = fields.Boolean(compute='_compute_cash_control', string='Has Cash Control')
@@ -98,7 +95,10 @@ class PosSession(models.Model):
     def write(self, vals):
         if vals.get('state') == 'closed':
             for record in self:
-                record.config_id._notify(('CLOSING_SESSION', {'login_number': self.env.context.get('login_number', False), 'session_id': record.id}))
+                record.config_id._notify(('CLOSING_SESSION', {
+                    'device_identifier': self.env.context.get('device_identifier', False),
+                    'session_id': record.id
+                }))
         return super().write(vals)
 
     @api.model
@@ -331,43 +331,6 @@ class PosSession(models.Model):
                 '\n'.join(f'{invoice.name} - {invoice.state}' for invoice in unposted_invoices)
             ))
 
-    def _create_sequences(self):
-        for session in self:
-            order_seq = self.env['ir.sequence'].sudo().create({
-                'name': _("PoS Order Sequence of Session %s", session.id),
-                'code': f'pos.order_{session.id}',
-            })
-            login_number_seq = self.env['ir.sequence'].sudo().create({
-                'name': _("Login Number Sequence of Session %s", session.id),
-                'code': f'pos.login_number_{session.id}',
-            })
-            session.write({
-                'order_seq_id': order_seq.id,
-                'login_number_seq_id': login_number_seq.id
-            })
-
-    def get_next_order_refs(self, login_number=0, ref_prefix=None, tracking_prefix=''):
-        """
-        Generates a consistent set of tracking_number, sequence_number and pos_reference for a new pos.order.
-        Side-effect: Calling this will increment the order_seq_id.
-        Convention: `login_number != 0` means the order is created in the classic PoS UI.
-            During self-ordering workflow (kiosk and mobile), login_number = 0 is used.
-        Returns:
-            (pos_reference: string, sequence_number: int, tracking_number: string)
-        """
-        self.ensure_one()
-
-        sequence_num = int(self.order_seq_id._next())
-
-        YY = fields.Datetime.now().strftime('%y')
-        LL = f"{login_number % 100:02}"
-        SSS = f"{self.id:03}"
-        F = 0  # -> means server-generated pos_reference
-        OOOO = f"{sequence_num:04}"
-        order_ref = f"{ref_prefix} {YY}{LL}-{SSS}-{F}{OOOO}" if ref_prefix else f"{YY}{LL}-{SSS}-{F}{OOOO}"
-
-        return order_ref, sequence_num, tracking_prefix + f"{sequence_num:03}"
-
     @api.model_create_multi
     def create(self, vals_list):
         for vals in vals_list:
@@ -393,21 +356,12 @@ class PosSession(models.Model):
         else:
             sessions = super().create(vals_list)
 
-        sessions._create_sequences()
         sessions.action_pos_session_open()
-
         return sessions
 
     def unlink(self):
         self.statement_line_ids.unlink()
-        sequences_to_unlink = (self.order_seq_id | self.login_number_seq_id)
-        result = super().unlink()
-        sequences_to_unlink.unlink()
-        return result
-
-    def login(self):
-        self.ensure_one()
-        return self.login_number_seq_id._next()
+        return super().unlink()
 
     def action_pos_session_open(self):
         # we only open sessions that haven't already been opened

--- a/addons/point_of_sale/models/res_config_settings.py
+++ b/addons/point_of_sale/models/res_config_settings.py
@@ -98,7 +98,6 @@ class ResConfigSettings(models.TransientModel):
     pos_rounding_method = fields.Many2one(related='pos_config_id.rounding_method', readonly=False)
     pos_route_id = fields.Many2one(related='pos_config_id.route_id', readonly=False)
     pos_selectable_categ_ids = fields.Many2many('pos.category', compute='_compute_pos_selectable_categ_ids')
-    pos_sequence_id = fields.Many2one(related='pos_config_id.sequence_id')
     pos_set_maximum_difference = fields.Boolean(related='pos_config_id.set_maximum_difference', readonly=False)
     pos_ship_later = fields.Boolean(related='pos_config_id.ship_later', readonly=False)
     pos_tax_regime_selection = fields.Boolean(related='pos_config_id.tax_regime_selection', readonly=False)
